In [103]:
!pip install -q condacolab
import condacolab
condacolab.install()

✨🍰✨ Everything looks OK!


In [104]:
!mamba create -n oldgensim python=3.7 "numpy<1.20" "scipy<1.6" -y
!conda run -n oldgensim pip install gensim==3.8.3


Looking for: ['python=3.7', "numpy[version='<1.20']", "scipy[version='<1.6']"]

[+] 0.0s
[+] 0.1s
conda-forge/linux-64   1%
conda-forge/noarch    ⣾  [+] 0.2s
conda-forge/linux-64   6%
conda-forge/noarch     8%[+] 0.3s
conda-forge/linux-64  13%
conda-forge/noarch    22%[+] 0.4s
conda-forge/linux-64  20%
conda-forge/noarch    37%[+] 0.5s
conda-forge/linux-64  23%
conda-forge/noarch    44%[+] 0.6s
conda-forge/linux-64  28%
conda-forge/noarch    54%[+] 0.7s
conda-forge/linux-64  32%
conda-forge/noarch    59%[+] 0.8s
conda-forge/linux-64  34%
conda-forge/noarch    66%[+] 0.9s
conda-forge/linux-64  34%
conda-forge/noarch    66%[+] 1.0s
conda-forge/linux-64  34%
conda-forge/noarch    66%[+] 1.1s
conda-forge/linux-64  34%
conda-forge/noarch    66%[+] 1.2s
conda-forge/linux-64  34%
conda-forge/noarch    66%[+] 1.3s
conda-forge/linux-64  34%
conda-forge/noarch    66%[+] 1.4s
conda-forge/linux-64  34%
conda-forge/noarch    69%[+] 1.5s
conda-forge/linux-64  34%
conda-forge/noarch    69%[+] 1.6s
c

In [105]:
convert_script = """
import gdown, zipfile
from gensim.models import Word2Vec

file_id = "0B0ZXk88koS2KbDhXdWg1Q2RydlU"
gdown.download(f"https://drive.google.com/uc?id={file_id}", "ko.zip", quiet=False)

with zipfile.ZipFile("ko.zip", "r") as z:
    z.extractall("./ko")

model = Word2Vec.load("./ko/ko.bin")
model.wv.save_word2vec_format("ko_word2vec_format.bin", binary=True)
print("변환 완료")
"""

with open("convert.py", "w") as f:
    f.write(convert_script)

!conda run -n oldgensim pip install gdown
!conda run -n oldgensim python convert.py

  Using cached gdown-4.7.3-py3-none-any.whl.metadata (4.4 kB)
  Using cached filelock-3.12.2-py3-none-any.whl.metadata (2.7 kB)
  Using cached requests-2.31.0-py3-none-any.whl.metadata (4.6 kB)
  Using cached tqdm-4.68.2-py3-none-any.whl.metadata (58 kB)
  Using cached beautifulsoup4-4.15.0-py3-none-any.whl.metadata (3.8 kB)
  Using cached soupsieve-2.4.1-py3-none-any.whl.metadata (4.7 kB)
  Using cached typing_extensions-4.7.1-py3-none-any.whl.metadata (3.1 kB)
  Using cached charset_normalizer-3.4.9-py3-none-any.whl.metadata (41 kB)
  Using cached idna-3.10-py3-none-any.whl.metadata (10 kB)
  Using cached urllib3-2.0.7-py3-none-any.whl.metadata (6.6 kB)
  Using cached certifi-2026.6.17-py3-none-any.whl.metadata (2.5 kB)
  Using cached PySocks-1.7.1-py3-none-any.whl.metadata (13 kB)
  Using cached importlib_metadata-6.7.0-py3-none-any.whl.metadata (4.9 kB)
  Using cached zipp-3.15.0-py3-none-any.whl.metadata (3.7 kB)
Using cached gdown-4.7.3-py3-none-any.whl (16 kB)
Using cached beaut

# **Step 1. 데이터 다운로드**

In [106]:
!pip install konlpy
!pip install mecab-python3
!pip install python-mecab-ko

  Using cached python_mecab_ko-1.3.7-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.4 kB)
Using cached python_mecab_ko-1.3.7-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (580 kB)


In [107]:
!pip uninstall -y konlpy mecab-python3 python-mecab-ko

Found existing installation: konlpy 0.6.0
Uninstalling konlpy-0.6.0:
  Successfully uninstalled konlpy-0.6.0
Found existing installation: mecab-python3 1.0.12
Uninstalling mecab-python3-1.0.12:
  Successfully uninstalled mecab-python3-1.0.12
Found existing installation: python-mecab-ko 1.3.7
Uninstalling python-mecab-ko-1.3.7:
  Successfully uninstalled python-mecab-ko-1.3.7


In [108]:
!git clone https://github.com/SOMJANG/Mecab-ko-for-Google-Colab.git
%cd Mecab-ko-for-Google-Colab
!bash install_mecab-ko_on_colab_light_220429.sh
%cd ..

fatal: destination path 'Mecab-ko-for-Google-Colab' already exists and is not an empty directory.
/content/Mecab-ko-for-Google-Colab
Installing konlpy.....
  Using cached konlpy-0.6.0-py2.py3-none-any.whl.metadata (1.9 kB)
Using cached konlpy-0.6.0-py2.py3-none-any.whl (19.4 MB)
Done
Installing mecab-0.996-ko-0.9.2.tar.gz.....
from https://bitbucket.org/eunjeon/mecab-ko/downloads/mecab-0.996-ko-0.9.2.tar.gz
--2026-07-15 10:18:57--  https://bitbucket.org/eunjeon/mecab-ko/downloads/mecab-0.996-ko-0.9.2.tar.gz
Resolving bitbucket.org (bitbucket.org)... 104.192.142.24, 104.192.142.26, 104.192.142.25, ...
Connecting to bitbucket.org (bitbucket.org)|104.192.142.24|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://bbuseruploads.s3.amazonaws.com/eunjeon/mecab-ko/downloads/mecab-0.996-ko-0.9.2.tar.gz?response-content-disposition=attachment%3B%20filename%3D%22mecab-0.996-ko-0.9.2.tar.gz%22&response-content-encoding=None&AWSAccessKeyId=ASIA6KOSE3BNJJ6YEN5K&Sig

In [109]:
from konlpy.tag import Mecab
mecab = Mecab()
print(mecab.morphs("안녕하세요 반갑습니다"))

['안녕', '하', '세요', '반갑', '습니다']


In [110]:
import re
import pandas as pd
from konlpy.tag import Mecab

# 데이터 다운로드

def load_data():
    """
    songys/Chatbot_data 저장소의 csv 파일을 읽어와
    questions, answers 리스트로 반환한다.
    """
    csv_url = "https://raw.githubusercontent.com/songys/Chatbot_data/master/ChatbotData.csv"
    data = pd.read_csv(csv_url)

    questions = data["Q"].tolist()
    answers = data["A"].tolist()

    return questions, answers

# **Step 2. 데이터 정제**

In [111]:
def preprocess_sentence(sentence):
    """
    - 영문자는 모두 소문자로 변환
    - 한글, 영문자, 숫자, 주요 특수문자(.,!?)를 제외한 문자는 모두 제거
    """
    sentence = sentence.lower().strip()
    sentence = re.sub(r"[^0-9a-zA-Z가-힣.,!?\s]", "", sentence)
    sentence = re.sub(r"\s+", " ", sentence).strip()
    return sentence

# **Step 3. 데이터 토큰화**

In [112]:
def build_corpus(src_data, tgt_data, tokenize_fn, max_len=40):
    """
    소스/타겟 문장을 정제 + 토큰화 + 길이 필터링 + 중복 제거하여 corpus 생성
    """
    src_corpus = []
    tgt_corpus = []

    seen_src = set()
    seen_tgt = set()

    for src_sentence, tgt_sentence in zip(src_data, tgt_data):
        src_clean = preprocess_sentence(src_sentence)
        tgt_clean = preprocess_sentence(tgt_sentence)

        src_tokens = tokenize_fn(src_clean)
        tgt_tokens = tokenize_fn(tgt_clean)

        if len(src_tokens) > max_len or len(tgt_tokens) > max_len:
            continue

        src_key = " ".join(src_tokens)
        tgt_key = " ".join(tgt_tokens)

        if src_key in seen_src or tgt_key in seen_tgt:
            continue

        seen_src.add(src_key)
        seen_tgt.add(tgt_key)

        src_corpus.append(src_tokens)
        tgt_corpus.append(tgt_tokens)

    return src_corpus, tgt_corpus


# 실행부

if __name__ == "__main__":
    questions, answers = load_data()
    print(f"질문 개수: {len(questions)}, 답변 개수: {len(answers)}")

    mecab = Mecab()
    que_corpus, ans_corpus = build_corpus(
        questions, answers, mecab.morphs, max_len=40
    )

    print(f"토큰화 및 필터링 후 데이터 개수: {len(que_corpus)}")
    print("샘플 질문 토큰:", que_corpus[0])
    print("샘플 답변 토큰:", ans_corpus[0])

질문 개수: 11823, 답변 개수: 11823
토큰화 및 필터링 후 데이터 개수: 7681
샘플 질문 토큰: ['12', '시', '땡', '!']
샘플 답변 토큰: ['하루', '가', '또', '가', '네요', '.']


# **Step 4. Augmentation(Lexical Substitution)**

In [113]:
!pip install --upgrade gensim
!pip install gdown

In [114]:
import random
from gensim.models import KeyedVectors


# 4-1. 변환된 word2vec 모델 로드

word2vec = KeyedVectors.load_word2vec_format("ko_word2vec_format.bin", binary=True)
print(word2vec.most_similar("강아지"))   # 정상 로드 확인용

# 4-2. Lexical Substitution 함수

def lexical_sub(tokens, word2vec):
    if len(tokens) == 0:
        return tokens

    indices = list(range(len(tokens)))
    random.shuffle(indices)

    for idx in indices:
        token = tokens[idx]
        try:
            similar_words = word2vec.most_similar(token)   # .wv 없이 바로 호출
        except KeyError:
            continue

        if not similar_words:
            continue

        new_word = random.choice(similar_words)[0]
        new_tokens = tokens[:idx] + [new_word] + tokens[idx + 1:]
        return new_tokens

    return tokens

# 4-3. corpus 전체에 augmentation 적용

def augment_corpus(corpus, word2vec):
    return [lexical_sub(tokens, word2vec) for tokens in corpus]


# 실행부

que_corpus_aug = augment_corpus(que_corpus, word2vec)
ans_corpus_aug = augment_corpus(ans_corpus, word2vec)

que_corpus_total = que_corpus + que_corpus_aug + que_corpus
ans_corpus_total = ans_corpus + ans_corpus + ans_corpus_aug

print(f"원본 데이터 개수 : {len(que_corpus)}")
print(f"augmentation 후 전체 데이터 개수 : {len(que_corpus_total)}")

print("\n[예시] 원본 질문 :", que_corpus[0])
print("[예시] augmentation 질문 :", que_corpus_aug[0])
print("[예시] 원본 답변 :", ans_corpus[0])
print("[예시] augmentation 답변 :", ans_corpus_aug[0])

[('고양이', 0.7290452718734741), ('거위', 0.7185635566711426), ('토끼', 0.7056224346160889), ('멧돼지', 0.6950401067733765), ('엄마', 0.6934333443641663), ('난쟁이', 0.6806551814079285), ('한마리', 0.677029550075531), ('아가씨', 0.6750352382659912), ('아빠', 0.6729634404182434), ('목걸이', 0.6512461304664612)]
원본 데이터 개수 : 7681
augmentation 후 전체 데이터 개수 : 23043

[예시] 원본 질문 : ['12', '시', '땡', '!']
[예시] augmentation 질문 : ['12', '시', '땡', '크레딧']
[예시] 원본 답변 : ['하루', '가', '또', '가', '네요', '.']
[예시] augmentation 답변 : ['하루', '가', '또', '로', '네요', '.']


# **Step 5. 데이터 벡터화**

In [115]:
import tensorflow as tf

# 5-1. 타겟 데이터에 <start>, <end> 토큰 추가

def add_start_end_token(corpus):
    """
    토큰화된 문장 리스트 각각의 맨 앞에 <start>, 맨 뒤에 <end> 토큰을 추가
    예: ["12", "시", "땡", "!"] -> ["<start>", "12", "시", "땡", "!", "<end>"]
    """
    return [["<start>"] + sentence + ["<end>"] for sentence in corpus]


# ans_corpus 완성 (타겟 데이터 전체에 special token 부착)
ans_corpus = add_start_end_token(ans_corpus_total)
que_corpus = que_corpus_total  # 소스 데이터는 그대로 사용

print("[확인] 타겟 샘플:", ans_corpus[0])


# 5-2. 단어 사전 구축 (que_corpus + ans_corpus 결합)

def build_tokenizer(que_corpus, ans_corpus):
    """
    que_corpus, ans_corpus를 합쳐 하나의 단어 사전을 만든다.
    filters=''  : 이미 형태소 단위로 토큰화되어 있으므로 별도 필터링 안 함
    lower=False : preprocess_sentence에서 이미 소문자 처리를 했으므로 중복 처리 방지
    oov_token   : 사전에 없는 단어 처리를 위한 토큰
    """
    tokenizer = tf.keras.preprocessing.text.Tokenizer(
        filters="",
        lower=False,
        oov_token="<unk>"
    )
    tokenizer.fit_on_texts(que_corpus + ans_corpus)
    return tokenizer


tokenizer = build_tokenizer(que_corpus, ans_corpus)

vocab_size = len(tokenizer.word_index) + 1
print(f"단어 사전 크기: {vocab_size}")


# 5-3. 벡터화 (정수 인코딩 + 패딩)

que_sequences = tokenizer.texts_to_sequences(que_corpus)
ans_sequences = tokenizer.texts_to_sequences(ans_corpus)

enc_train = tf.keras.preprocessing.sequence.pad_sequences(
    que_sequences, padding="post"
)
dec_train = tf.keras.preprocessing.sequence.pad_sequences(
    ans_sequences, padding="post"
)

print(f"enc_train shape: {enc_train.shape}")
print(f"dec_train shape: {dec_train.shape}")

print("\n[예시] 원본 질문 토큰   :", que_corpus[0])
print("[예시] enc_train 벡터   :", enc_train[0])
print("[예시] 원본 답변 토큰   :", ans_corpus[0])
print("[예시] dec_train 벡터   :", dec_train[0])

[확인] 타겟 샘플: ['<start>', '하루', '가', '또', '가', '네요', '.', '<end>']
단어 사전 크기: 8471
enc_train shape: (23043, 32)
dec_train shape: (23043, 42)

[예시] 원본 질문 토큰   : ['12', '시', '땡', '!']
[예시] enc_train 벡터   : [2314  210 2953  112    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0]
[예시] 원본 답변 토큰   : ['<start>', '하루', '가', '또', '가', '네요', '.', '<end>']
[예시] dec_train 벡터   : [  3 278   9 139   9  42   2   4   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0]


# **Step 6. 모델학습**

Positional Encoding

In [116]:
# Positional Encoding 구현
def positional_encoding(pos, d_model):
    def cal_angle(position, i):
        return position / np.power(10000, (2*(i//2)) / np.float32(d_model))

    def get_posi_angle_vec(position):
        return [cal_angle(position, i) for i in range(d_model)]

    sinusoid_table = np.array([get_posi_angle_vec(pos_i) for pos_i in range(pos)])

    sinusoid_table[:, 0::2] = np.sin(sinusoid_table[:, 0::2])
    sinusoid_table[:, 1::2] = np.cos(sinusoid_table[:, 1::2])

    return sinusoid_table
print("슝=3")

슝=3


마스크 생성

In [117]:
import torch

def generate_padding_mask(seq: torch.Tensor) -> torch.Tensor:
    """
    seq: shape [batch_size, seq_len]의 입력 (토큰 ID 텐서)
    반환: shape [batch_size, 1, 1, seq_len]의 패딩 마스크
         (seq == 0)인 위치가 1, 나머지는 0
    """
    # (seq == 0)은 불리언 텐서를 반환 -> float()로 형변환 -> (1.0 or 0.0)
    # 차원 확장: [batch_size, seq_len] → [batch_size, 1, 1, seq_len]
    return (seq == 0).unsqueeze(1).unsqueeze(2).float()


def generate_lookahead_mask(size: int) -> torch.Tensor:
    """
    size: 문장(시퀀스) 길이
    반환: shape [size, size],
         i < j (대각선 위)에 해당하는 위치가 1, 아닌 곳은 0
         (미래 토큰을 가리기 위한 마스크)
    """
    # triu(diagonal=1)은 주대각선 위가 1, 아래가 0인 텐서를 만들어 줌
    return torch.triu(torch.ones(size, size), diagonal=1)


def generate_masks(src: torch.Tensor, tgt: torch.Tensor):
    """
    src, tgt: shape [batch_size, seq_len]
    3가지 마스크를 반환:
      - enc_mask: 인코더 입력용 패딩 마스크
      - dec_enc_mask: 디코더-인코더 어텐션용 패딩 마스크
      - dec_mask: 디코더 자기어텐션용 마스크(룩어헤드 + 패딩)

    각각의 shape:
      - enc_mask, dec_enc_mask: [batch_size, 1, 1, src_seq_len]
      - dec_mask: [batch_size, 1, tgt_seq_len, tgt_seq_len]
    """
    # 1) 인코더 입력용 패딩 마스크
    enc_mask = generate_padding_mask(src)
    # 2) 디코더에서 인코더 값을 볼 때 사용하는 마스크 (src 마스크 재사용)
    dec_enc_mask = generate_padding_mask(src)

    # 3) 디코더 자기어텐션 마스크 (미래 토큰 방지 룩어헤드 + tgt 자체 패딩 마스크)
    dec_lookahead_mask = generate_lookahead_mask(tgt.shape[1])  # [tgt_seq_len, tgt_seq_len]
    dec_tgt_padding_mask = generate_padding_mask(tgt)           # [batch_size, 1, 1, tgt_seq_len]

    # 룩어헤드 마스크를 (batch 차원과 head 차원을 가상으로) 확장
    dec_lookahead_mask = dec_lookahead_mask.unsqueeze(0).unsqueeze(1)  # [1, 1, seq_len, seq_len]

    # 패딩 + 룩어헤드 마스크 병합
    # 브로드캐스팅에 의해 shape [batch_size, 1, tgt_seq_len, tgt_seq_len]이 됨

    dec_tgt_padding_mask = dec_tgt_padding_mask.to(device)
    dec_lookahead_mask = dec_lookahead_mask.to(device)

    dec_mask = torch.max(dec_tgt_padding_mask, dec_lookahead_mask)

    return enc_mask, dec_enc_mask, dec_mask

print("슝=3")

슝=3


Multi-head Attention

In [118]:
import torch
import torch.nn as nn
import torch.nn.functional as F

class MultiHeadAttention(nn.Module):
    def __init__(self, d_model, num_heads):
        super(MultiHeadAttention, self).__init__()
        self.num_heads = num_heads
        self.d_model = d_model

        # d_model을 num_heads로 나눈 만큼이 각 head가 담당할 차원 수
        self.depth = d_model // num_heads

        # Query, Key, Value를 구하는 선형 레이어
        self.W_q = nn.Linear(d_model, d_model)
        self.W_k = nn.Linear(d_model, d_model)
        self.W_v = nn.Linear(d_model, d_model)

        # 최종적으로 head들의 출력을 결합해주는 선형 레이어
        self.linear = nn.Linear(d_model, d_model)

    def scaled_dot_product_attention(self, Q, K, V, mask=None):
        """
        Q, K, V:  [batch_size, num_heads, seq_len, depth]
        mask:     [batch_size, 1, seq_len, seq_len] 혹은
                  [batch_size, num_heads, seq_len, seq_len]
                  (어텐션에서 제외할 위치=1, 사용할 위치=0)
        """
        # d_k = depth
        d_k = Q.size(-1)  # K.shape[-1]도 동일
        # Q와 K의 전치 곱: (batch_size, num_heads, seq_len, seq_len)
        QK = torch.matmul(Q, K.transpose(-1, -2))

        # 스케일링
        scaled_qk = QK / torch.sqrt(torch.tensor(d_k, dtype=torch.float32))

        # 마스크가 있는 경우 -1e9(매우 작은 수)를 더하여 softmax 후 확률이 0에 가깝도록 처리
        if mask is not None:
            scaled_qk = scaled_qk + (mask * -1e9)

        attentions = F.softmax(scaled_qk, dim=-1)  # (batch_size, num_heads, seq_len, seq_len)
        out = torch.matmul(attentions, V)         # (batch_size, num_heads, seq_len, depth)

        return out, attentions

    def split_heads(self, x):
        """
        x: [batch_size, seq_len, d_model]
        반환: [batch_size, num_heads, seq_len, depth]
        """
        bsz, seq_len, _ = x.size()
        # d_model -> (num_heads * depth)이므로 view로 재배치
        x = x.view(bsz, seq_len, self.num_heads, self.depth)
        # (batch_size, seq_len, num_heads, depth) -> (batch_size, num_heads, seq_len, depth)
        x = x.permute(0, 2, 1, 3)
        return x

    def combine_heads(self, x):
        """
        x: [batch_size, num_heads, seq_len, depth]
        반환: [batch_size, seq_len, d_model]
        """
        bsz, num_heads, seq_len, depth = x.size()
        # (batch_size, num_heads, seq_len, depth) -> (batch_size, seq_len, num_heads, depth)
        x = x.permute(0, 2, 1, 3).contiguous()
        x = x.view(bsz, seq_len, self.d_model)
        return x

    def forward(self, Q, K, V, mask=None):
        """
        Q, K, V: [batch_size, seq_len, d_model]
        mask:    [batch_size, 1, seq_len, seq_len] 혹은
                 [batch_size, num_heads, seq_len, seq_len]
        """
        # W_q, W_k, W_v는 각각 (d_model -> d_model) 선형 변환
        WQ = self.W_q(Q)  # [batch_size, seq_len, d_model]
        WK = self.W_k(K)  # [batch_size, seq_len, d_model]
        WV = self.W_v(V)  # [batch_size, seq_len, d_model]

        # 멀티헤드 분할
        WQ_splits = self.split_heads(WQ)  # [batch_size, num_heads, seq_len, depth]
        WK_splits = self.split_heads(WK)
        WV_splits = self.split_heads(WV)

        # Scaled dot-product attention
        out, attention_weights = self.scaled_dot_product_attention(
            WQ_splits, WK_splits, WV_splits, mask
        )

        # head 결과 결합 후 최종 선형
        out = self.combine_heads(out)  # [batch_size, seq_len, d_model]
        out = self.linear(out)         # [batch_size, seq_len, d_model]

        return out, attention_weights

print("슝=3")

슝=3


Position-wise Feed Forward Network

In [119]:
class PoswiseFeedForwardNet(nn.Module):
    def __init__(self, d_model, d_ff):
        super(PoswiseFeedForwardNet, self).__init__()
        self.d_model = d_model
        self.d_ff = d_ff

        self.fc1 = nn.Linear(d_model, d_ff)
        self.fc2 = nn.Linear(d_ff, d_model)
        self.relu = nn.ReLU()

    def forward(self, x):
        out = self.relu(self.fc1(x))  # 첫 번째 Dense + ReLU
        out = self.fc2(out)          # 두 번째 Dense
        return out

print("슝=3")

슝=3


Encoder Layer

In [120]:
class EncoderLayer(nn.Module):
    def __init__(self, d_model, n_heads, d_ff, dropout):
        super(EncoderLayer, self).__init__()
        self.enc_self_attn = MultiHeadAttention(d_model, n_heads)
        self.ffn = PoswiseFeedForwardNet(d_model, d_ff)

        # nn.LayerNorm은 마지막 차원(d_model)을 기준으로 정규화
        self.norm_1 = nn.LayerNorm(d_model, eps=1e-6)
        self.norm_2 = nn.LayerNorm(d_model, eps=1e-6)

        self.do = nn.Dropout(dropout)

    def forward(self, x, mask):
        # Multi-Head Attention 단계
        residual = x
        out = self.norm_1(x)
        out, enc_attn = self.enc_self_attn(out, out, out, mask)
        out = self.do(out)
        out = out + residual  # residual connection

        # Position-Wise Feed Forward 단계
        residual = out
        out = self.norm_2(out)
        out = self.ffn(out)
        out = self.do(out)
        out = out + residual  # residual connection

        return out, enc_attn

print("슝=3")

슝=3


Decoder Layer

In [121]:
class DecoderLayer(nn.Module):
    def __init__(self, d_model, num_heads, d_ff, dropout):
        super(DecoderLayer, self).__init__()
        self.dec_self_attn = MultiHeadAttention(d_model, num_heads)
        self.enc_dec_attn = MultiHeadAttention(d_model, num_heads)
        self.ffn = PoswiseFeedForwardNet(d_model, d_ff)

        self.norm_1 = nn.LayerNorm(d_model, eps=1e-6)
        self.norm_2 = nn.LayerNorm(d_model, eps=1e-6)
        self.norm_3 = nn.LayerNorm(d_model, eps=1e-6)

        self.do = nn.Dropout(dropout)

    def forward(self, x, enc_out, dec_enc_mask, padding_mask):
        # Masked Multi-Head Attention
        residual = x
        out = self.norm_1(x)
        out, dec_attn = self.dec_self_attn(out, out, out, mask=padding_mask)
        out = self.do(out)
        out = out + residual

        # Encoder-Decoder Multi-Head Attention (주의: Q, K, V 순서)
        residual = out
        out = self.norm_2(out)
        out, dec_enc_attn = self.enc_dec_attn(out, enc_out, enc_out, mask=dec_enc_mask)
        out = self.do(out)
        out = out + residual

        # Position-Wise Feed Forward Network
        residual = out
        out = self.norm_3(out)
        out = self.ffn(out)
        out = self.do(out)
        out = out + residual

        return out, dec_attn, dec_enc_attn

print("슝=3")

슝=3


Encoder

In [122]:
class Encoder(nn.Module):
    def __init__(self, n_layers, d_model, n_heads, d_ff, dropout):
        super(Encoder, self).__init__()
        self.n_layers = n_layers
        self.enc_layers = nn.ModuleList(
            [EncoderLayer(d_model, n_heads, d_ff, dropout) for _ in range(n_layers)]
        )
        self.do = nn.Dropout(dropout)  # 필요 시 입력에 dropout 적용 가능

    def forward(self, x, mask):
        out = x
        enc_attns = []
        for i in range(self.n_layers):
            out, enc_attn = self.enc_layers[i](out, mask)
            enc_attns.append(enc_attn)
        return out, enc_attns

# 사용 예시: Encoder 인스턴스 생성 후 forward 호출
# encoder = Encoder(n_layers, d_model, n_heads, d_ff, dropout)
# out, enc_attns = encoder(x, mask)
print("슝=3")

슝=3


Decoder

In [123]:
class Decoder(nn.Module):
    def __init__(self, n_layers, d_model, n_heads, d_ff, dropout):
        super(Decoder, self).__init__()
        self.n_layers = n_layers
        self.dec_layers = nn.ModuleList(
            [DecoderLayer(d_model, n_heads, d_ff, dropout) for _ in range(n_layers)]
        )

    def forward(self, x, enc_out, dec_enc_mask, padding_mask):
        out = x
        dec_attns = []
        dec_enc_attns = []
        for i in range(self.n_layers):
            out, dec_attn, dec_enc_attn = self.dec_layers[i](out, enc_out, dec_enc_mask, padding_mask)
            dec_attns.append(dec_attn)
            dec_enc_attns.append(dec_enc_attn)
        return out, dec_attns, dec_enc_attns

print("슝=3")

슝=3


Transformer 전체 모델 조립

In [124]:
import math

class Transformer(nn.Module):
    def __init__(self, n_layers, d_model, n_heads, d_ff,
                 src_vocab_size, tgt_vocab_size, pos_len,
                 dropout=0.2, shared_fc=True, shared_emb=False):
        super(Transformer, self).__init__()
        self.d_model = float(d_model)

        if shared_emb:
            self.enc_emb = self.dec_emb = nn.Embedding(src_vocab_size, d_model)
        else:
            self.enc_emb = nn.Embedding(src_vocab_size, d_model)
            self.dec_emb = nn.Embedding(tgt_vocab_size, d_model)

        pos_encoding_np = positional_encoding(pos_len, d_model)
        self.register_buffer("pos_encoding", torch.tensor(pos_encoding_np, dtype=torch.float32))

        self.do = nn.Dropout(dropout)

        self.encoder = Encoder(n_layers, d_model, n_heads, d_ff, dropout)
        self.decoder = Decoder(n_layers, d_model, n_heads, d_ff, dropout)

        self.fc = nn.Linear(d_model, tgt_vocab_size)

        self.shared_fc = shared_fc
        if shared_fc:
            self.fc.weight = self.dec_emb.weight

        self._init_weights()   # ★ 추가된 부분

    def _init_weights(self):   # ★ 새로 추가된 메서드
        nn.init.normal_(self.dec_emb.weight, mean=0.0, std=self.d_model ** -0.5)
        if self.enc_emb is not self.dec_emb:
            nn.init.normal_(self.enc_emb.weight, mean=0.0, std=self.d_model ** -0.5)
        for name, p in self.named_parameters():
            if p.dim() > 1 and "emb" not in name:
                nn.init.xavier_uniform_(p)

    def embedding(self, emb, x):
        seq_len = x.size(1)
        out = emb(x)
        if self.shared_fc:
            out = out * math.sqrt(self.d_model)
        out = out + self.pos_encoding[:seq_len, :].unsqueeze(0)
        out = self.do(out)
        return out

    def forward(self, enc_in, dec_in, enc_mask, dec_enc_mask, dec_mask):
        enc_in_emb = self.embedding(self.enc_emb, enc_in)
        dec_in_emb = self.embedding(self.dec_emb, dec_in)

        enc_out, enc_attns = self.encoder(enc_in_emb, enc_mask)
        dec_out, dec_attns, dec_enc_attns = self.decoder(dec_in_emb, enc_out, dec_enc_mask, dec_mask)

        logits = self.fc(dec_out)
        return logits, enc_attns, dec_attns, dec_enc_attns

print("Transformer 클래스 재정의 완료")

Transformer 클래스 재정의 완료


In [125]:
class Encoder(nn.Module):
    def __init__(self, n_layers, d_model, n_heads, d_ff, dropout):
        super(Encoder, self).__init__()
        self.n_layers = n_layers
        self.enc_layers = nn.ModuleList(
            [EncoderLayer(d_model, n_heads, d_ff, dropout) for _ in range(n_layers)]
        )
        self.do = nn.Dropout(dropout)
        self.final_norm = nn.LayerNorm(d_model, eps=1e-6)   # ← 추가

    def forward(self, x, mask):
        out = x
        enc_attns = []
        for i in range(self.n_layers):
            out, enc_attn = self.enc_layers[i](out, mask)
            enc_attns.append(enc_attn)
        out = self.final_norm(out)   # ← 추가
        return out, enc_attns

In [126]:
class Decoder(nn.Module):
    def __init__(self, n_layers, d_model, n_heads, d_ff, dropout):
        super(Decoder, self).__init__()
        self.n_layers = n_layers
        self.dec_layers = nn.ModuleList(
            [DecoderLayer(d_model, n_heads, d_ff, dropout) for _ in range(n_layers)]
        )
        self.final_norm = nn.LayerNorm(d_model, eps=1e-6)   # ← 추가

    def forward(self, x, enc_out, dec_enc_mask, padding_mask):
        out = x
        dec_attns = []
        dec_enc_attns = []
        for i in range(self.n_layers):
            out, dec_attn, dec_enc_attn = self.dec_layers[i](out, enc_out, dec_enc_mask, padding_mask)
            dec_attns.append(dec_attn)
            dec_enc_attns.append(dec_enc_attn)
        out = self.final_norm(out)   # ← 추가
        return out, dec_attns, dec_enc_attns

모델 인스턴스 생성

In [127]:
import numpy as np

#  하이퍼파라미터 설정

VOCAB_SIZE = len(tokenizer.word_index) + 1
print("VOCAB_SIZE:", VOCAB_SIZE)

N_LAYERS = 2
D_MODEL = 256          # 512 -> 256으로 축소
N_HEADS = 4             # 256 / 8 = 32, 정수로 나누어떨어지므로 그대로 유지
D_FF = 1024             # d_model을 절반으로 줄였으니 d_ff도 절반(2048 -> 1024)으로 같이 축소 권장
DROPOUT = 0.3

# 모델, optimizer 초기화
transformer = Transformer(
    n_layers=N_LAYERS, d_model=D_MODEL, n_heads=N_HEADS, d_ff=D_FF,
    src_vocab_size=VOCAB_SIZE, tgt_vocab_size=VOCAB_SIZE,
    pos_len=50, dropout=DROPOUT, shared_fc=True, shared_emb=True
)
transformer = transformer.to(device)

learning_rate = LearningRateScheduler(D_MODEL, warmup_steps=900)
optimizer = torch.optim.AdamW(transformer.parameters(), lr=learning_rate(1), betas=(0.9, 0.98), eps=1e-9, weight_decay=0.01)

VOCAB_SIZE: 8471


Learning Rate Scheduler

In [128]:
class LearningRateScheduler:
    def __init__(self, d_model, warmup_steps=900):
        self.d_model = d_model
        self.warmup_steps = warmup_steps

    def __call__(self, step):
        # step을 float으로 변환하여 지수 연산이 제대로 수행되도록 함
        step = float(step)
        arg1 = step ** -0.5
        arg2 = step * (self.warmup_steps ** -1.5)
        return (self.d_model ** -0.5) * min(arg1, arg2)

print("슝=3")

슝=3


Loss Function 정의

In [129]:
def loss_function(real, pred, smoothing=0.1):
    real = real.to(device)
    pred = pred.to(device)

    num_classes = pred.size(-1)
    log_probs = F.log_softmax(pred, dim=-1)

    true_dist = torch.zeros_like(log_probs).scatter_(-1, real.unsqueeze(-1), 1.0)
    true_dist = true_dist * (1 - smoothing) + smoothing / num_classes

    loss_ = -(true_dist * log_probs).sum(dim=-1)

    mask = (real != 0).float()
    loss_ = loss_ * mask

    return loss_.sum() / mask.sum()

Train Step 정의

In [130]:
def train_step(src, tgt, model, optimizer):
    model.train()  # 모델을 training 모드로 전환
    optimizer.zero_grad()

    # tgt의 오른쪽 시프트: decoder input과 gold target 분리
    tgt_in = tgt[:, :-1]  # Decoder의 입력
    gold = tgt[:, 1:]     # Decoder의 정답(target)

    # 마스크 생성 (generate_masks는 PyTorch용으로 변환된 함수여야 합니다)
    enc_mask, dec_enc_mask, dec_mask = generate_masks(src, tgt_in)

    src = src.to(device)
    tgt_in = tgt_in.to(device)
    enc_mask = enc_mask.to(device)
    dec_enc_mask = dec_enc_mask.to(device)
    dec_mask = dec_mask.to(device)
    gold = gold.to(device)

    # 모델 forward pass
    predictions, enc_attns, dec_attns, dec_enc_attns = model(src, tgt_in, enc_mask, dec_enc_mask, dec_mask)
    print("logit range:", predictions.min().item(), predictions.max().item())

    # loss 계산
    loss = loss_function(gold, predictions)

    # 역전파 수행 및 파라미터 업데이트
    loss.backward()
    optimizer.step()

    return loss, enc_attns, dec_attns, dec_enc_attns

print("슝=3")

슝=3


모델학습

In [131]:
from torch.utils.data import TensorDataset, DataLoader

BATCH_SIZE = 64

enc_train_tensor = torch.tensor(enc_train, dtype=torch.long)
dec_train_tensor = torch.tensor(dec_train, dtype=torch.long)

train_dataset = TensorDataset(enc_train_tensor, dec_train_tensor)
train_dataloader = DataLoader(
    train_dataset,
    batch_size=BATCH_SIZE,
    shuffle=True,
    drop_last=True
)

print("steps_per_epoch:", len(train_dataloader))

steps_per_epoch: 360


In [132]:
from tqdm.notebook import tqdm

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("사용 device:", device)

사용 device: cuda


In [133]:
%%time

EPOCHS = 30
global_step = 0   # ★ 추가: 루프 시작 전, for epoch 밖에서 한 번만 선언

for epoch in range(EPOCHS):
    total_loss = 0.0
    dataset_count = len(train_dataloader)
    tqdm_bar = tqdm(total=dataset_count)

    for batch, (src, tgt) in enumerate(train_dataloader):
        global_step += 1                              # ★ 추가: 매 스텝마다 증가
        current_lr = learning_rate(global_step)        # ★ 추가: 현재 스텝의 lr 계산
        for param_group in optimizer.param_groups:      # ★ 추가: optimizer에 반영
            param_group['lr'] = current_lr

        loss, enc_attns, dec_attns, dec_enc_attns = train_step(src, tgt, transformer, optimizer)

        total_loss += loss.item()
        tqdm_bar.set_postfix({"Batch Loss": f"{loss.item():.4f}", "lr": f"{current_lr:.2e}"})  # lr도 같이 표시
        tqdm_bar.update(1)

    tqdm_bar.close()
    print(f"Epoch {epoch+1}, Loss: {total_loss / dataset_count:.4f}, lr: {current_lr:.2e}")

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -5.575188159942627 10.3492431640625
logit range: -5.447021961212158 10.587076187133789
logit range: -5.401340007781982 10.450688362121582
logit range: -6.150663375854492 10.007518768310547
logit range: -5.147103309631348 10.200061798095703
logit range: -5.433114051818848 10.214725494384766
logit range: -5.373763084411621 10.944303512573242
logit range: -5.256121635437012 10.37817096710205
logit range: -5.177740097045898 10.093899726867676
logit range: -5.479682922363281 10.279277801513672
logit range: -5.52913761138916 10.006878852844238
logit range: -5.381733417510986 10.204797744750977
logit range: -5.425323486328125 10.740968704223633
logit range: -5.366839408874512 10.37637996673584
logit range: -5.301875114440918 9.732763290405273
logit range: -5.0162858963012695 9.77301025390625
logit range: -5.7404680252075195 9.590472221374512
logit range: -5.278421878814697 9.99374771118164
logit range: -5.464752197265625 9.677056312561035
logit range: -5.2640862464904785 10.11663

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -6.175123691558838 12.224547386169434
logit range: -6.110162258148193 12.502174377441406
logit range: -6.483755111694336 12.010900497436523
logit range: -6.2186784744262695 11.94821834564209
logit range: -6.373146057128906 11.83215618133545
logit range: -6.331015586853027 11.688375473022461
logit range: -6.210869312286377 11.756909370422363
logit range: -6.305843830108643 11.74360466003418
logit range: -6.205411911010742 11.670292854309082
logit range: -6.203254222869873 11.76657485961914
logit range: -6.370162010192871 12.210992813110352
logit range: -6.205708980560303 11.924455642700195
logit range: -6.596273422241211 12.291031837463379
logit range: -6.2425103187561035 11.808734893798828
logit range: -6.701853275299072 12.024941444396973
logit range: -6.285964488983154 12.222904205322266
logit range: -6.412929534912109 12.361412048339844
logit range: -6.3072004318237305 12.13770580291748
logit range: -6.7136712074279785 12.092769622802734
logit range: -6.430752754211426 

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -6.6322550773620605 12.095255851745605
logit range: -7.013251304626465 12.118942260742188
logit range: -7.098954200744629 11.960700035095215
logit range: -7.122293949127197 12.219122886657715
logit range: -7.35627555847168 11.881816864013672
logit range: -7.331890106201172 11.727922439575195
logit range: -7.504395484924316 12.025640487670898
logit range: -7.22120475769043 12.230823516845703
logit range: -7.449862957000732 12.344242095947266
logit range: -7.221690654754639 12.200490951538086
logit range: -7.076900959014893 12.028725624084473
logit range: -7.07637882232666 12.217870712280273
logit range: -7.153992652893066 12.210912704467773
logit range: -7.651273727416992 12.13947582244873
logit range: -7.0768890380859375 12.211506843566895
logit range: -7.499761581420898 12.17480182647705
logit range: -6.98963737487793 12.01738452911377
logit range: -7.14616060256958 12.220467567443848
logit range: -6.849796772003174 12.050346374511719
logit range: -7.130759239196777 12.30

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -8.34537124633789 12.619668960571289
logit range: -8.939230918884277 12.67955207824707
logit range: -8.316924095153809 12.49229907989502
logit range: -8.461164474487305 12.62936782836914
logit range: -9.844940185546875 12.387748718261719
logit range: -8.783488273620605 12.21190357208252
logit range: -8.570154190063477 11.952689170837402
logit range: -8.540703773498535 11.969145774841309
logit range: -8.624123573303223 12.354994773864746
logit range: -8.45659065246582 12.096380233764648
logit range: -8.227482795715332 12.28466510772705
logit range: -8.493691444396973 12.495952606201172
logit range: -8.52038860321045 12.35147762298584
logit range: -8.693431854248047 12.581793785095215
logit range: -9.410993576049805 13.42950439453125
logit range: -8.335468292236328 13.005990982055664
logit range: -9.478503227233887 13.055495262145996
logit range: -8.622603416442871 12.271957397460938
logit range: -8.045759201049805 12.524194717407227
logit range: -8.772505760192871 12.529266

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -9.537897109985352 12.54835319519043
logit range: -9.591331481933594 12.984259605407715
logit range: -9.456201553344727 12.933490753173828
logit range: -9.26101303100586 12.935117721557617
logit range: -9.455520629882812 13.12073802947998
logit range: -9.2709321975708 13.465951919555664
logit range: -8.966618537902832 13.49051570892334
logit range: -9.305583953857422 12.8214111328125
logit range: -8.998624801635742 12.60576343536377
logit range: -9.079344749450684 12.665452003479004
logit range: -9.320661544799805 12.40666675567627
logit range: -9.119836807250977 12.645737648010254
logit range: -9.55893325805664 12.912215232849121
logit range: -9.58555793762207 13.089913368225098
logit range: -9.8333740234375 13.083295822143555
logit range: -10.491520881652832 12.97952938079834
logit range: -9.30873966217041 13.4600830078125
logit range: -9.636238098144531 13.677377700805664
logit range: -9.761246681213379 13.303522109985352
logit range: -9.700535774230957 13.8787183761596

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -10.556220054626465 13.262672424316406
logit range: -10.454967498779297 13.845681190490723
logit range: -9.902745246887207 13.25901985168457
logit range: -10.15278434753418 12.82833480834961
logit range: -9.752184867858887 13.764741897583008
logit range: -9.871232986450195 13.971519470214844
logit range: -10.122443199157715 13.561651229858398
logit range: -10.006699562072754 12.770073890686035
logit range: -10.546388626098633 13.565935134887695
logit range: -9.993012428283691 13.74992847442627
logit range: -10.134109497070312 13.508835792541504
logit range: -10.356456756591797 14.157172203063965
logit range: -10.695734977722168 13.752336502075195
logit range: -11.110583305358887 14.13782787322998
logit range: -10.191290855407715 14.651422500610352
logit range: -9.962006568908691 14.064061164855957
logit range: -9.857879638671875 14.12816333770752
logit range: -10.077418327331543 13.475406646728516
logit range: -11.20356559753418 14.35009479522705
logit range: -10.155744552

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -11.217756271362305 14.210112571716309
logit range: -10.346611976623535 13.939834594726562
logit range: -10.529704093933105 14.163309097290039
logit range: -11.243099212646484 14.204137802124023
logit range: -10.955371856689453 14.010059356689453
logit range: -11.010063171386719 14.24079704284668
logit range: -11.199923515319824 14.561433792114258
logit range: -11.799930572509766 14.327776908874512
logit range: -11.431241989135742 13.977209091186523
logit range: -10.616843223571777 13.874519348144531
logit range: -11.296390533447266 13.960212707519531
logit range: -10.078583717346191 14.045453071594238
logit range: -10.931614875793457 13.765287399291992
logit range: -10.969194412231445 13.7286958694458
logit range: -11.51313591003418 13.802868843078613
logit range: -10.687034606933594 14.434577941894531
logit range: -10.5737886428833 14.06716251373291
logit range: -11.345723152160645 14.161330223083496
logit range: -11.384300231933594 14.034345626831055
logit range: -11.74

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -11.414216995239258 14.721440315246582
logit range: -12.02768325805664 14.41163444519043
logit range: -12.207415580749512 14.181207656860352
logit range: -11.728755950927734 14.005455017089844
logit range: -11.840188980102539 14.495137214660645
logit range: -12.811838150024414 14.899773597717285
logit range: -11.497437477111816 15.152774810791016
logit range: -12.261494636535645 14.337313652038574
logit range: -11.397037506103516 14.873147010803223
logit range: -11.819920539855957 13.92620849609375
logit range: -11.40937614440918 14.65280532836914
logit range: -11.393885612487793 14.7523775100708
logit range: -11.292550086975098 15.48857307434082
logit range: -11.462475776672363 14.685173988342285
logit range: -11.589890480041504 14.832063674926758
logit range: -11.781564712524414 15.055681228637695
logit range: -12.075919151306152 14.602608680725098
logit range: -11.976780891418457 14.05035400390625
logit range: -11.287341117858887 14.036245346069336
logit range: -12.1694

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -12.718908309936523 15.180511474609375
logit range: -12.753514289855957 15.640196800231934
logit range: -11.539534568786621 15.449719429016113
logit range: -11.590224266052246 15.035778999328613
logit range: -11.737842559814453 14.369844436645508
logit range: -11.665472030639648 14.216740608215332
logit range: -11.606646537780762 14.717020988464355
logit range: -12.394780158996582 16.297035217285156
logit range: -11.46138858795166 15.434915542602539
logit range: -11.317270278930664 15.647286415100098
logit range: -12.10843276977539 15.15511417388916
logit range: -12.993803024291992 14.767400741577148
logit range: -12.04796314239502 15.107934951782227
logit range: -12.146744728088379 15.080228805541992
logit range: -12.457590103149414 14.775115013122559
logit range: -11.216181755065918 14.968145370483398
logit range: -12.32128620147705 15.0055570602417
logit range: -11.427618026733398 14.34589958190918
logit range: -12.261011123657227 15.339628219604492
logit range: -12.360

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -13.090291023254395 15.82972526550293
logit range: -12.468398094177246 14.81678295135498
logit range: -12.5115966796875 15.493632316589355
logit range: -12.481428146362305 15.573813438415527
logit range: -12.232730865478516 15.387104988098145
logit range: -12.134737968444824 14.737577438354492
logit range: -13.584994316101074 15.440193176269531
logit range: -12.517765998840332 15.107078552246094
logit range: -13.012347221374512 15.262804985046387
logit range: -12.647359848022461 15.823542594909668
logit range: -11.991026878356934 15.449766159057617
logit range: -12.021992683410645 15.329487800598145
logit range: -12.207118034362793 15.41392993927002
logit range: -12.630575180053711 16.34663200378418
logit range: -12.106202125549316 15.447957992553711
logit range: -12.238814353942871 14.896594047546387
logit range: -12.903630256652832 14.831768989562988
logit range: -12.053064346313477 15.426729202270508
logit range: -12.37588882446289 15.434815406799316
logit range: -12.03

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -13.56718921661377 14.962284088134766
logit range: -12.818056106567383 16.477643966674805
logit range: -12.273137092590332 14.813420295715332
logit range: -13.682130813598633 15.784859657287598
logit range: -14.522422790527344 15.700748443603516
logit range: -13.01632308959961 14.986356735229492
logit range: -12.512267112731934 15.234830856323242
logit range: -14.069544792175293 17.73134994506836
logit range: -13.819140434265137 16.755674362182617
logit range: -12.71104621887207 15.546515464782715
logit range: -12.22744083404541 16.11838722229004
logit range: -12.58496379852295 15.22826862335205
logit range: -12.741293907165527 14.654088020324707
logit range: -13.778213500976562 14.9946870803833
logit range: -12.714536666870117 15.612253189086914
logit range: -12.925236701965332 16.49306869506836
logit range: -15.288501739501953 15.825736045837402
logit range: -12.307209968566895 15.53446102142334
logit range: -12.626446723937988 16.69217300415039
logit range: -12.31999969

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -13.647518157958984 16.080408096313477
logit range: -12.759931564331055 15.214534759521484
logit range: -12.937810897827148 15.973304748535156
logit range: -12.861078262329102 15.52237319946289
logit range: -14.714110374450684 16.565237045288086
logit range: -13.067998886108398 16.96201515197754
logit range: -13.093040466308594 16.28756332397461
logit range: -13.572948455810547 15.18734359741211
logit range: -12.99909496307373 16.07577133178711
logit range: -13.472404479980469 15.403728485107422
logit range: -14.299581527709961 15.62031364440918
logit range: -13.448863983154297 15.30463981628418
logit range: -13.262249946594238 15.459169387817383
logit range: -13.078622817993164 15.806520462036133
logit range: -12.502875328063965 16.234052658081055
logit range: -13.389545440673828 14.742905616760254
logit range: -13.273122787475586 15.711257934570312
logit range: -12.969551086425781 16.019039154052734
logit range: -13.503568649291992 15.588501930236816
logit range: -13.384

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -12.88975715637207 15.81551456451416
logit range: -14.945725440979004 16.834442138671875
logit range: -13.861287117004395 16.779937744140625
logit range: -13.792596817016602 18.183256149291992
logit range: -13.528509140014648 16.451770782470703
logit range: -13.400211334228516 16.007240295410156
logit range: -12.814929008483887 16.543100357055664
logit range: -13.705960273742676 16.311798095703125
logit range: -13.60402774810791 15.202177047729492
logit range: -13.625658988952637 15.028565406799316
logit range: -13.343424797058105 16.569992065429688
logit range: -13.18175983428955 16.450714111328125
logit range: -13.529194831848145 16.164770126342773
logit range: -14.204516410827637 16.20639419555664
logit range: -13.269187927246094 16.53643226623535
logit range: -14.380436897277832 16.142549514770508
logit range: -13.642047882080078 16.91244888305664
logit range: -14.301606178283691 15.973912239074707
logit range: -14.141868591308594 15.458757400512695
logit range: -13.48

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -13.481887817382812 15.589487075805664
logit range: -14.032536506652832 16.677330017089844
logit range: -13.78376579284668 17.18649673461914
logit range: -14.277822494506836 15.339345932006836
logit range: -13.7789945602417 17.35310935974121
logit range: -14.225447654724121 17.722063064575195
logit range: -15.067477226257324 15.629217147827148
logit range: -13.68536376953125 16.30203628540039
logit range: -14.16118335723877 16.171031951904297
logit range: -13.614680290222168 15.56628704071045
logit range: -15.366193771362305 15.578191757202148
logit range: -15.062417030334473 15.689836502075195
logit range: -13.727171897888184 16.732295989990234
logit range: -13.714426040649414 16.158584594726562
logit range: -14.251667976379395 16.25065040588379
logit range: -15.320170402526855 17.200721740722656
logit range: -15.470060348510742 15.756585121154785
logit range: -13.451712608337402 15.96905517578125
logit range: -13.911508560180664 17.084260940551758
logit range: -13.247248

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -15.336838722229004 16.750673294067383
logit range: -13.53561782836914 16.045162200927734
logit range: -14.02396297454834 17.078968048095703
logit range: -14.234935760498047 16.920753479003906
logit range: -13.996682167053223 16.564533233642578
logit range: -14.44644546508789 16.135845184326172
logit range: -13.95602035522461 17.73538589477539
logit range: -14.478083610534668 16.28099822998047
logit range: -13.852254867553711 18.02000617980957
logit range: -14.761690139770508 16.214643478393555
logit range: -14.909299850463867 15.750028610229492
logit range: -14.111151695251465 16.994890213012695
logit range: -13.829158782958984 17.532447814941406
logit range: -14.176721572875977 15.492453575134277
logit range: -14.290623664855957 16.16791534423828
logit range: -13.948390007019043 16.15802764892578
logit range: -14.890596389770508 15.691368103027344
logit range: -14.068233489990234 15.639703750610352
logit range: -14.542428016662598 16.46953582763672
logit range: -15.96693

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -15.495918273925781 17.725719451904297
logit range: -14.462363243103027 16.027721405029297
logit range: -14.86801815032959 17.1240291595459
logit range: -14.611150741577148 16.23369789123535
logit range: -15.624796867370605 15.134642601013184
logit range: -15.513238906860352 16.12436294555664
logit range: -14.431228637695312 17.111722946166992
logit range: -13.930082321166992 16.0710506439209
logit range: -14.651552200317383 16.60470199584961
logit range: -16.20404624938965 16.45153045654297
logit range: -14.420047760009766 16.693078994750977
logit range: -14.555530548095703 16.555452346801758
logit range: -14.709497451782227 15.770101547241211
logit range: -15.001636505126953 16.559749603271484
logit range: -15.60912036895752 16.914243698120117
logit range: -15.243382453918457 16.902843475341797
logit range: -15.075188636779785 16.236740112304688
logit range: -14.261299133300781 16.581283569335938
logit range: -14.315150260925293 16.572650909423828
logit range: -14.228112

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -15.377423286437988 15.96326732635498
logit range: -15.246384620666504 16.801137924194336
logit range: -14.644728660583496 15.797802925109863
logit range: -14.976428985595703 16.45388412475586
logit range: -15.533782005310059 16.01211166381836
logit range: -14.077404022216797 16.755117416381836
logit range: -14.805699348449707 17.490060806274414
logit range: -14.827409744262695 17.548913955688477
logit range: -14.775121688842773 17.948486328125
logit range: -14.969529151916504 16.793601989746094
logit range: -15.737595558166504 16.850322723388672
logit range: -14.680421829223633 16.338769912719727
logit range: -14.549057006835938 16.279743194580078
logit range: -14.265182495117188 16.216705322265625
logit range: -14.803314208984375 16.12144660949707
logit range: -14.09401798248291 17.494993209838867
logit range: -14.107051849365234 15.434351921081543
logit range: -14.844077110290527 16.27075958251953
logit range: -15.089635848999023 16.799692153930664
logit range: -15.0431

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -15.513738632202148 16.483732223510742
logit range: -14.824941635131836 17.681793212890625
logit range: -15.15866470336914 15.59166431427002
logit range: -15.077934265136719 17.121782302856445
logit range: -14.923449516296387 16.435575485229492
logit range: -15.23974609375 16.400880813598633
logit range: -14.901893615722656 16.675928115844727
logit range: -15.142990112304688 16.898971557617188
logit range: -15.078841209411621 18.543991088867188
logit range: -15.637234687805176 17.543302536010742
logit range: -15.642908096313477 16.50469398498535
logit range: -14.705248832702637 16.21820068359375
logit range: -14.7938814163208 16.529911041259766
logit range: -14.630606651306152 16.351133346557617
logit range: -14.799851417541504 16.908000946044922
logit range: -15.129583358764648 16.438514709472656
logit range: -14.9476318359375 16.682329177856445
logit range: -15.06811237335205 16.279176712036133
logit range: -15.089487075805664 16.16908073425293
logit range: -15.263338088

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -16.001304626464844 16.333953857421875
logit range: -14.705989837646484 16.817651748657227
logit range: -14.798873901367188 16.234600067138672
logit range: -15.170819282531738 16.033878326416016
logit range: -15.666800498962402 16.47617530822754
logit range: -15.685617446899414 15.81206226348877
logit range: -14.917978286743164 17.615232467651367
logit range: -15.842032432556152 17.50442886352539
logit range: -15.293240547180176 17.025989532470703
logit range: -15.734503746032715 16.128374099731445
logit range: -15.611454010009766 17.216840744018555
logit range: -16.772981643676758 16.804685592651367
logit range: -15.544281959533691 17.968368530273438
logit range: -16.071819305419922 17.523622512817383
logit range: -14.343116760253906 17.154582977294922
logit range: -14.85220718383789 16.171796798706055
logit range: -15.40997314453125 17.61368751525879
logit range: -15.011857032775879 16.835538864135742
logit range: -15.232440948486328 17.158309936523438
logit range: -15.9

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -15.942405700683594 17.343782424926758
logit range: -16.229825973510742 17.567249298095703
logit range: -15.261378288269043 16.047704696655273
logit range: -15.258852005004883 16.309528350830078
logit range: -15.796453475952148 16.61006736755371
logit range: -16.111377716064453 17.06612777709961
logit range: -16.098678588867188 17.894123077392578
logit range: -14.838983535766602 16.21340560913086
logit range: -15.648017883300781 16.675752639770508
logit range: -14.965670585632324 16.803476333618164
logit range: -15.565998077392578 16.6768741607666
logit range: -16.186988830566406 18.58946990966797
logit range: -14.957240104675293 16.105539321899414
logit range: -15.485666275024414 16.546886444091797
logit range: -15.439461708068848 17.16828155517578
logit range: -15.414884567260742 17.813621520996094
logit range: -16.307003021240234 16.33040428161621
logit range: -15.088647842407227 16.187307357788086
logit range: -15.789743423461914 16.697763442993164
logit range: -15.732

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -17.86488151550293 16.913137435913086
logit range: -15.317843437194824 15.806855201721191
logit range: -15.71264362335205 16.304733276367188
logit range: -15.614351272583008 17.16217803955078
logit range: -15.354164123535156 16.22011947631836
logit range: -16.055471420288086 16.495492935180664
logit range: -14.95002269744873 18.01189613342285
logit range: -16.267454147338867 16.33467674255371
logit range: -15.407739639282227 16.050567626953125
logit range: -14.968023300170898 15.86091136932373
logit range: -16.04935073852539 16.200531005859375
logit range: -17.823177337646484 17.812868118286133
logit range: -16.483882904052734 16.461130142211914
logit range: -17.26911735534668 18.189146041870117
logit range: -16.456769943237305 17.86871337890625
logit range: -15.752057075500488 15.805765151977539
logit range: -16.116777420043945 16.874034881591797
logit range: -16.385644912719727 17.73647689819336
logit range: -15.496499061584473 15.858430862426758
logit range: -16.8841686

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -16.145029067993164 16.20027732849121
logit range: -16.320323944091797 16.834720611572266
logit range: -15.814535140991211 15.55721378326416
logit range: -15.696511268615723 16.82331085205078
logit range: -16.372020721435547 15.815800666809082
logit range: -17.076738357543945 16.368988037109375
logit range: -14.87410831451416 16.97197914123535
logit range: -15.387859344482422 16.86390495300293
logit range: -17.097970962524414 16.498258590698242
logit range: -17.434307098388672 17.2356014251709
logit range: -15.697334289550781 15.577738761901855
logit range: -17.031248092651367 17.59798240661621
logit range: -15.615023612976074 17.839170455932617
logit range: -15.998296737670898 16.822032928466797
logit range: -16.051557540893555 17.701152801513672
logit range: -16.85152816772461 17.247770309448242
logit range: -15.937051773071289 15.678116798400879
logit range: -15.458151817321777 17.512250900268555
logit range: -16.081024169921875 16.971813201904297
logit range: -15.18421

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -19.133121490478516 16.93621826171875
logit range: -16.13198471069336 17.202028274536133
logit range: -17.277734756469727 15.572623252868652
logit range: -16.453594207763672 16.392248153686523
logit range: -15.5796480178833 16.90941047668457
logit range: -15.966917037963867 17.456872940063477
logit range: -16.188640594482422 17.672874450683594
logit range: -16.448057174682617 16.542089462280273
logit range: -15.868110656738281 15.920947074890137
logit range: -16.226835250854492 15.885611534118652
logit range: -15.71821117401123 17.401321411132812
logit range: -17.343698501586914 16.97956085205078
logit range: -16.223905563354492 15.708205223083496
logit range: -17.440353393554688 17.665653228759766
logit range: -15.447405815124512 16.36488151550293
logit range: -15.572351455688477 16.003509521484375
logit range: -14.953357696533203 17.418312072753906
logit range: -15.064773559570312 16.897123336791992
logit range: -16.918392181396484 16.59457778930664
logit range: -16.1795

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -15.714929580688477 15.70305061340332
logit range: -15.6882905960083 16.82856559753418
logit range: -16.581676483154297 17.461490631103516
logit range: -17.23972511291504 16.411861419677734
logit range: -17.288454055786133 16.442659378051758
logit range: -17.058353424072266 16.96331214904785
logit range: -16.024023056030273 16.465755462646484
logit range: -15.947334289550781 16.608232498168945
logit range: -16.264379501342773 17.666736602783203
logit range: -16.286239624023438 16.561546325683594
logit range: -17.05039405822754 16.69266700744629
logit range: -17.053730010986328 16.846126556396484
logit range: -16.56593132019043 16.37526512145996
logit range: -16.5512752532959 18.661102294921875
logit range: -16.819759368896484 17.332611083984375
logit range: -16.69269561767578 17.115482330322266
logit range: -16.678327560424805 16.460519790649414
logit range: -16.14948844909668 16.449871063232422
logit range: -16.317548751831055 17.088064193725586
logit range: -17.407831192

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -16.644453048706055 18.423442840576172
logit range: -17.08466911315918 16.567201614379883
logit range: -17.656734466552734 16.368608474731445
logit range: -16.41431427001953 16.28900909423828
logit range: -15.955390930175781 16.439035415649414
logit range: -15.63798713684082 17.308368682861328
logit range: -16.815799713134766 17.114303588867188
logit range: -16.109081268310547 16.65458869934082
logit range: -17.36252784729004 16.07339859008789
logit range: -16.216035842895508 16.23028564453125
logit range: -16.723417282104492 16.661758422851562
logit range: -15.68038558959961 16.162067413330078
logit range: -16.1041202545166 16.630048751831055
logit range: -16.257728576660156 16.440507888793945
logit range: -19.185123443603516 16.465158462524414
logit range: -16.43649673461914 16.366113662719727
logit range: -17.484621047973633 16.37002182006836
logit range: -17.327404022216797 16.085525512695312
logit range: -15.883133888244629 16.78988265991211
logit range: -17.728321075

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -16.974178314208984 16.708751678466797
logit range: -17.04831886291504 16.250511169433594
logit range: -17.77667999267578 16.308795928955078
logit range: -18.11823081970215 16.113466262817383
logit range: -16.33222770690918 16.59882926940918
logit range: -16.174434661865234 16.741037368774414
logit range: -16.52045249938965 17.186378479003906
logit range: -17.281307220458984 16.498519897460938
logit range: -15.814718246459961 16.61623191833496
logit range: -17.883853912353516 16.71762466430664
logit range: -17.470447540283203 17.222341537475586
logit range: -16.989456176757812 16.943674087524414
logit range: -16.511098861694336 17.128841400146484
logit range: -16.37397575378418 16.516908645629883
logit range: -17.069272994995117 16.636377334594727
logit range: -17.341236114501953 15.90097427368164
logit range: -18.493070602416992 15.687384605407715
logit range: -16.98036766052246 16.04230499267578
logit range: -17.075359344482422 18.291860580444336
logit range: -15.9485359

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -17.032068252563477 16.83460807800293
logit range: -18.164770126342773 16.056854248046875
logit range: -16.544382095336914 16.46100425720215
logit range: -17.44693374633789 16.69788360595703
logit range: -17.281967163085938 16.959260940551758
logit range: -17.216487884521484 15.774940490722656
logit range: -17.615503311157227 16.371788024902344
logit range: -16.611616134643555 17.09333038330078
logit range: -17.080705642700195 16.966176986694336
logit range: -16.45337677001953 17.676748275756836
logit range: -19.00157356262207 17.192487716674805
logit range: -16.83561897277832 15.87092399597168
logit range: -18.08193588256836 17.936756134033203
logit range: -17.756240844726562 18.9184627532959
logit range: -17.749900817871094 15.974475860595703
logit range: -17.778112411499023 15.674922943115234
logit range: -16.064435958862305 16.264162063598633
logit range: -17.570661544799805 15.847535133361816
logit range: -17.858545303344727 16.342424392700195
logit range: -16.6797504

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -16.299327850341797 16.919898986816406
logit range: -17.189607620239258 15.773390769958496
logit range: -18.85719871520996 15.338888168334961
logit range: -17.23817253112793 15.923336029052734
logit range: -16.943374633789062 15.31507396697998
logit range: -17.289785385131836 15.76280689239502
logit range: -17.976530075073242 16.08646583557129
logit range: -16.671052932739258 17.08308982849121
logit range: -15.742959022521973 16.05815315246582
logit range: -17.786130905151367 17.54716682434082
logit range: -17.569374084472656 17.54332733154297
logit range: -17.641834259033203 18.1000919342041
logit range: -18.193544387817383 16.469940185546875
logit range: -16.95355224609375 17.78314781188965
logit range: -18.715347290039062 16.61858367919922
logit range: -16.929292678833008 15.819769859313965
logit range: -16.08677864074707 15.405942916870117
logit range: -16.055530548095703 15.74135684967041
logit range: -17.6474609375 16.7806339263916
logit range: -16.425512313842773 15

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -17.49116325378418 16.680030822753906
logit range: -16.62798500061035 16.629032135009766
logit range: -17.719629287719727 16.38482093811035
logit range: -18.402109146118164 15.668608665466309
logit range: -16.93805694580078 15.489312171936035
logit range: -17.869306564331055 16.305835723876953
logit range: -16.524147033691406 15.4578218460083
logit range: -16.73529624938965 15.887527465820312
logit range: -18.967405319213867 16.07610321044922
logit range: -17.747512817382812 18.666027069091797
logit range: -17.062606811523438 16.52700424194336
logit range: -16.94098472595215 16.944700241088867
logit range: -17.16895294189453 16.153339385986328
logit range: -16.922290802001953 15.359718322753906
logit range: -16.471960067749023 17.051542282104492
logit range: -17.667217254638672 16.96580696105957
logit range: -17.484333038330078 17.976545333862305
logit range: -17.00592803955078 15.860223770141602
logit range: -17.533397674560547 18.069292068481445
logit range: -17.69610214

  0%|          | 0/360 [00:00<?, ?it/s]

logit range: -16.795166015625 15.786421775817871
logit range: -17.39842987060547 16.21393394470215
logit range: -18.085779190063477 17.953853607177734
logit range: -16.483732223510742 17.960641860961914
logit range: -17.938093185424805 15.23837661743164
logit range: -16.547027587890625 15.506597518920898
logit range: -18.096660614013672 16.046552658081055
logit range: -17.414348602294922 15.587392807006836
logit range: -16.452035903930664 15.967326164245605
logit range: -17.07848358154297 15.529071807861328
logit range: -17.695371627807617 16.45477867126465
logit range: -16.738248825073242 15.33098030090332
logit range: -16.750503540039062 16.221532821655273
logit range: -19.93990707397461 16.012248992919922
logit range: -19.940275192260742 16.479894638061523
logit range: -17.898149490356445 15.450279235839844
logit range: -17.20947265625 15.661035537719727
logit range: -16.8989200592041 15.494790077209473
logit range: -17.286842346191406 18.319116592407227
logit range: -17.48545455932

예문에 대한 답변 생성

In [141]:
def translate(sentence, model, tokenizer, max_len=40):
    model.eval()
    tokens = mecab.morphs(preprocess_sentence(sentence))
    input_seq = tokenizer.texts_to_sequences([tokens])
    enc_in = torch.tensor(input_seq, dtype=torch.long).to(device)

    dec_in = torch.tensor([[tokenizer.word_index["<start>"]]], dtype=torch.long).to(device)

    with torch.no_grad():
        for _ in range(max_len):
            enc_mask, dec_enc_mask, dec_mask = generate_masks(enc_in, dec_in)
            logits, *_ = model(enc_in, dec_in, enc_mask, dec_enc_mask, dec_mask)
            next_token = logits[:, -1, :].argmax(dim=-1).unsqueeze(0)
            dec_in = torch.cat([dec_in, next_token], dim=1)
            if next_token.item() == tokenizer.word_index["<end>"]:
                break

    result_tokens = [tokenizer.index_word[idx.item()] for idx in dec_in[0][1:]]
    return " ".join(result_tokens)

In [144]:
# 예문에 대한 답변 생성

def translate(sentence, model, tokenizer, max_len=40):
    model.eval()
    tokens = mecab.morphs(preprocess_sentence(sentence))
    input_seq = tokenizer.texts_to_sequences([tokens])
    enc_in = torch.tensor(input_seq, dtype=torch.long).to(device)

    dec_in = torch.tensor([[tokenizer.word_index["<start>"]]], dtype=torch.long).to(device)

    with torch.no_grad():
        for _ in range(max_len):
            enc_mask, dec_enc_mask, dec_mask = generate_masks(enc_in, dec_in)
            logits, *_ = model(enc_in, dec_in, enc_mask, dec_enc_mask, dec_mask)
            next_token = logits[:, -1, :].argmax(dim=-1).unsqueeze(0)
            dec_in = torch.cat([dec_in, next_token], dim=1)
            if next_token.item() == tokenizer.word_index["<end>"]:
                break

    result_tokens = [tokenizer.index_word[idx.item()] for idx in dec_in[0][1:]]
    return " ".join(result_tokens)


# 예문 정의

example_sentences = [
    "지루하다, 놀러가고 싶어.",
    "오늘 일찍 일어났더니 피곤하다.",
    "간만에 여자친구랑 데이트 하기로 했어.",
    "집에 있는다는 소리야.",
]


# 답변 생성

translations = [translate(sent, transformer, tokenizer) for sent in example_sentences]


# 제출 형식 출력

print("# 예문")
for i, sent in enumerate(example_sentences, start=1):
    print(f"{i}. {sent}")

print("\n---\n")
print("# 제출\n")

print("Translations")
for i, trans in enumerate(translations, start=1):
    print(f"> {i}. {trans}")

print("\nHyperparameters")
print(f"> n_layers: {N_LAYERS}")
print(f"> d_model: {D_MODEL}")
print(f"> n_heads: {N_HEADS}")
print(f"> d_ff: {D_FF}")
print(f"> dropout: {DROPOUT}")

print("\nTraining Parameters")
print(f"> Warmup Steps: {learning_rate.warmup_steps}")
print(f"> Batch Size: {BATCH_SIZE}")
print(f"> Epoch At: {EPOCHS}")

# 예문
1. 지루하다, 놀러가고 싶어.
2. 오늘 일찍 일어났더니 피곤하다.
3. 간만에 여자친구랑 데이트 하기로 했어.
4. 집에 있는다는 소리야.

---

# 제출

Translations
> 1. 다른 생각 을 해 보 세요 . <end>
> 2. 이제 일어날 때 좋 은 추억 이 에요 . <end>
> 3. 떨리 는 감정 은 그 자체 로 소중 하 죠 . <end>
> 4. 집 에 들어가 뜨거운 물 어 보 세요 . <end>

Hyperparameters
> n_layers: 2
> d_model: 256
> n_heads: 4
> d_ff: 1024
> dropout: 0.3

Training Parameters
> Warmup Steps: 900
> Batch Size: 64
> Epoch At: 30


- 예문1과 3정도는 위로와 공감 정도의 답변으로 괜찮으나 예문 2와 4의 답변은 동문서답임   

# **Step 7. 성능 측정**

In [146]:
# BLEU Score 계산

import random
from nltk.translate.bleu_score import sentence_bleu, corpus_bleu, SmoothingFunction


def generate_prediction_tokens(sentence_tokens, model, tokenizer, max_len=40):
    """
    이미 토큰화되어 있는 소스 문장(형태소 리스트)을 입력받아
    모델이 생성한 답변을 토큰 리스트로 반환 (greedy decoding)
    """
    model.eval()
    input_seq = tokenizer.texts_to_sequences([sentence_tokens])
    enc_in = torch.tensor(input_seq, dtype=torch.long).to(device)

    dec_in = torch.tensor([[tokenizer.word_index["<start>"]]], dtype=torch.long).to(device)

    with torch.no_grad():
        for _ in range(max_len):
            enc_mask, dec_enc_mask, dec_mask = generate_masks(enc_in, dec_in)
            logits, *_ = model(enc_in, dec_in, enc_mask, dec_enc_mask, dec_mask)
            next_token = logits[:, -1, :].argmax(dim=-1).unsqueeze(0)
            dec_in = torch.cat([dec_in, next_token], dim=1)
            if next_token.item() == tokenizer.word_index["<end>"]:
                break

    result_tokens = [
        tokenizer.index_word[idx.item()] for idx in dec_in[0][1:]
        if idx.item() != tokenizer.word_index["<end>"]
    ]
    return result_tokens


# 평가용 샘플 추출 (전체 다 돌리면 오래 걸리므로 일부만 샘플링 권장)

NUM_SAMPLES = 300   # 평가에 사용할 샘플 수 (필요시 조정)
random.seed(42)
sample_indices = random.sample(range(len(que_corpus)), min(NUM_SAMPLES, len(que_corpus)))

references = []   # 정답 토큰 리스트들의 리스트 (문장당 참조가 1개이므로 [[ref_tokens]] 형태)
hypotheses = []    # 모델이 생성한 토큰 리스트

for idx in sample_indices:
    src_tokens = que_corpus[idx]
    ref_tokens = [tok for tok in ans_corpus[idx] if tok not in ("<start>", "<end>")]

    pred_tokens = generate_prediction_tokens(src_tokens, transformer, tokenizer)

    references.append([ref_tokens])   # corpus_bleu는 참조를 리스트의 리스트로 받음
    hypotheses.append(pred_tokens)


# BLEU Score 계산 (문장이 짧아서 smoothing 적용 권장)

smoothie = SmoothingFunction().method4

bleu1 = corpus_bleu(references, hypotheses, weights=(1, 0, 0, 0), smoothing_function=smoothie)
bleu2 = corpus_bleu(references, hypotheses, weights=(0.5, 0.5, 0, 0), smoothing_function=smoothie)
bleu3 = corpus_bleu(references, hypotheses, weights=(0.33, 0.33, 0.33, 0), smoothing_function=smoothie)
bleu4 = corpus_bleu(references, hypotheses, weights=(0.25, 0.25, 0.25, 0.25), smoothing_function=smoothie)

print(f"BLEU-1: {bleu1:.4f}")
print(f"BLEU-2: {bleu2:.4f}")
print(f"BLEU-3: {bleu3:.4f}")
print(f"BLEU-4: {bleu4:.4f}")


BLEU-1: 0.9024
BLEU-2: 0.8780
BLEU-3: 0.8569
BLEU-4: 0.8340


- BLEU 점수(0.83~0.90)는 실제 Chatbot 답변결과와 비교하면 지나치게 높다고 생각됨
- 모델이 학습 데이터의 패턴을 암기해서 높은 점수가 나오지 않았나 의심이 되어 8개의 새로운 예시문과 답변을 작성하셔 추가 테스트

In [149]:

# Held-out 평가: 학습 데이터에 없는 새 문장 10개로 BLEU 계산

from nltk.translate.bleu_score import sentence_bleu, corpus_bleu, SmoothingFunction

smoothie = SmoothingFunction().method4

eval_pairs = [
    ("오늘 날씨가 너무 덥다.",                  "시원한 물이라도 한 잔 드세요."),
    ("몇일간 과로를 했더니 피곤하다.",          "푹 쉬세요."),
    ("요즘 기분이 계속 가라 앉아.",             "기분 전환할 수 있는 가벼운 운동을 해보세요."),
    ("회사 일이 너무 힘들어.",                  "고생 많으셨어요."),
    ("남자친구랑 헤어졌어.",                    "많이 힘드시겠어요."),
    ("이직을 고민하고 있어.",                   "신중하게 생각해보세요"),
    ("내일 발표가 있는데 너무 떨려",            "잘 하실 거예요."),
    ("혼자 있는 시간이 좋아졌어.",               "그것도 좋은 거예요."),
]


# 2) 정답(참조)과 모델 예측을 각각 토큰화

references = []
hypotheses = []
predictions_text = []

for question, answer in eval_pairs:
    # 정답도 동일한 전처리+토큰화 파이프라인을 거쳐야 공정한 비교가 됩니다
    ref_tokens = mecab.morphs(preprocess_sentence(answer))

    # 모델이 실제로 생성한 답변
    q_tokens = mecab.morphs(preprocess_sentence(question))
    pred_tokens = generate_prediction_tokens(q_tokens, transformer, tokenizer)

    references.append([ref_tokens])   # corpus_bleu는 [[ref]] 형태를 요구
    hypotheses.append(pred_tokens)
    predictions_text.append(" ".join(pred_tokens))


# 3) BLEU score 계산

bleu1 = corpus_bleu(references, hypotheses, weights=(1, 0, 0, 0), smoothing_function=smoothie)
bleu2 = corpus_bleu(references, hypotheses, weights=(0.5, 0.5, 0, 0), smoothing_function=smoothie)
bleu3 = corpus_bleu(references, hypotheses, weights=(0.33, 0.33, 0.33, 0), smoothing_function=smoothie)
bleu4 = corpus_bleu(references, hypotheses, weights=(0.25, 0.25, 0.25, 0.25), smoothing_function=smoothie)

print(f"BLEU-1: {bleu1:.4f}")
print(f"BLEU-2: {bleu2:.4f}")
print(f"BLEU-3: {bleu3:.4f}")
print(f"BLEU-4: {bleu4:.4f}")



# 4) 문장별 상세 비교

print("\n[문장별 비교]")
for i, (question, answer) in enumerate(eval_pairs):
    sent_bleu4 = sentence_bleu(references[i], hypotheses[i], weights=(0.25, 0.25, 0.25, 0.25), smoothing_function=smoothie)
    print(f"{i+1}. 질문   : {question}")
    print(f"   정답   : {answer}")
    print(f"   모델답변: {predictions_text[i]}")
    print(f"   BLEU-4 : {sent_bleu4:.4f}")
    print("-" * 40)

BLEU-1: 0.3492
BLEU-2: 0.2760
BLEU-3: 0.2384
BLEU-4: 0.2019

[문장별 비교]
1. 질문   : 오늘 날씨가 너무 덥다.
   정답   : 시원한 물이라도 한 잔 드세요.
   모델답변: 시원 한 물 이 라도 한 잔 드세요
   BLEU-4 : 0.8825
----------------------------------------
2. 질문   : 몇일간 과로를 했더니 피곤하다.
   정답   : 푹 쉬세요.
   모델답변: 정신 적 이 생길 거 예요 .
   BLEU-4 : 0.0324
----------------------------------------
3. 질문   : 요즘 기분이 계속 가라 앉아.
   정답   : 기분 전환할 수 있는 가벼운 운동을 해보세요.
   모델답변: 제 가 수면 위 로 올라갈 수 있 게 도와 드릴게요 .
   BLEU-4 : 0.0486
----------------------------------------
4. 질문   : 회사 일이 너무 힘들어.
   정답   : 고생 많으셨어요.
   모델답변: 회사 는 스트레스 의 원인 이 죠 .
   BLEU-4 : 0.0286
----------------------------------------
5. 질문   : 남자친구랑 헤어졌어.
   정답   : 많이 힘드시겠어요.
   모델답변: 힘들 겠 어요 .
   BLEU-4 : 0.2617
----------------------------------------
6. 질문   : 이직을 고민하고 있어.
   정답   : 신중하게 생각해보세요
   모델답변: 다니 면서 준비 해 보 세요 .
   BLEU-4 : 0.1931
----------------------------------------
7. 질문   : 내일 발표가 있는데 너무 떨려
   정답   : 잘 하실 거예요.
   모델답변: 더 많이 연습 하 고 준비 해 보 세요 .
   BLEU-4 : 0.0279
--------

- 예시문장 1번(0.88), 5번(0.26), 4번(0.03)에 대해서 답변을 보면 가고 다만 5번 경우는 다소 BLEU점수가 낮아 보임.
- 학습 데이터의 BlEU(0.83)는 높지만 held-out 문장의 BLEU(0.20)이 이 모델 본래의 성능 점수라고 보여짐.

-  학습 단계에서 logit이 수천 단위로 폭주하는 문제가 있었고 하나씩 점검해가며 Encoder/Decoder 최종 출력에 LayerNorm이 빠져있던 것을 확인.
- 이후에도 러닝레이트 스케줄이 실제로 업데이트되지 않아 학습이 멈춘 문제까지 발견했고 이를 통해 코드가 에러 없이 돌아간다고 모델이 실제로 학습되고 있는건 아니라는 사실을 알게 됨.